### PR TITLE
Added native sleep payloads

### PR DESCRIPTION
**Pull request by federicodotta**
Added native sleep payloads to payload generation and RMI exploit

### CODE REVIEW

**Comment by frohoff** [comment]
Thanks. This is certainly an interesting idea, though unfortunately a sleep() call may not be possible with all gadget chains. I'll leave this open for now as I have some refactors in flight and want to figure out if/how to fold this functionality in cleanly.

**Comment by federicodotta** [comment]
Hi frohoff! Thanks for the reply. Yes, sure! I will work on my branch for now and eventually we will merge the code in future.

I have added the sleep gadgets because I have implemented a scanner for the deserialization (plugin for Burp Suite) and the sleep function was the best way I founded for the detection. I tried other solutions (sleep in Runtime.exec, ping and DNS resolution, other commands in the Runtime.exec, etc.) but none of them are suitable for the detection (the sleep in Runtime.exec is asyncronous, DNS resolution sometimes is filtered, other commands in the Runtime.exec do not allow to easy understand if the command has worked or not because usually return no output).

I have implemented the sleep gadget for the Apache Commons Collections 3, Apache Commons Collections 4 and Spring but, as you said, I can't find a way to implement the sleep with Groovy. In Groovy chain, if I understand it well, you use the String.execute() method of Groovy that takes no arguments. I succeeded in lauch the Thread.sleep method but I can't find a way to pass to the method an argument (the time of the sleep). If you have some advice for me, they are very welcome! :)

Thank you a lot for you great tool and congratulations for your great discovery!

File: src/main/java/ysoserial/secmgr/ExecCheckingSecurityManager.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+public class ExecCheckingSecurityManager extends SecurityManager {
+	public ExecCheckingSecurityManager() {
+		this(true);
+	}
+
+	public ExecCheckingSecurityManager(boolean throwException) {
+		this.throwException = throwException;
+	}
+
+	private final boolean throwException;
```
Comment (frohoff) [inline src/main/java/ysoserial/secmgr/ExecCheckingSecurityManager.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://www.codacy.com/app/frohoff/ysoserial/file/2167893690/issues/source?bid=3161887&fileBranchId=3161890#l18)

File: src/main/java/ysoserial/secmgr/ExecCheckingSecurityManager.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+		SecurityManager sm = System.getSecurityManager(); // save sm
+		System.setSecurityManager(this);
+		try {
+			T result = callable.call();
+			if (throwException && ! getCmds().isEmpty()) {
+				throw new ExecException(getCmds().get(0));
+			}
+			return result;
+		} catch (Exception e) {
+			if (! (e instanceof ExecException) && throwException && ! getCmds().isEmpty()) {
```
Comment (frohoff) [inline src/main/java/ysoserial/secmgr/ExecCheckingSecurityManager.java]: Codacy Issue found: [An instanceof check is being performed on the caught exception.  Create a separate catch clause for this exception type.](https://www.codacy.com/app/frohoff/ysoserial/file/2167893690/issues/source?bid=3161887&fileBranchId=3161890#l78)

File: src/main/java/ysoserial/secmgr/ExecCheckingSecurityManager.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+		this(true);
+	}
+
+	public ExecCheckingSecurityManager(boolean throwException) {
+		this.throwException = throwException;
+	}
+
+	private final boolean throwException;
+
+	private final List<String> cmds = new LinkedList<String>();
```
Comment (frohoff) [inline src/main/java/ysoserial/secmgr/ExecCheckingSecurityManager.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://www.codacy.com/app/frohoff/ysoserial/file/2167893690/issues/source?bid=3161887&fileBranchId=3161890#l20)

File: src/main/java/ysoserial/secmgr/ExecCheckingSecurityManager.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+	public List<String> getCmds() {
+		return Collections.unmodifiableList(cmds);
+	}
+
+	@Override
+	public void checkPermission(final Permission perm) { }
```
Comment (frohoff) [inline src/main/java/ysoserial/secmgr/ExecCheckingSecurityManager.java]: Codacy Issue found: [Document empty method body](https://www.codacy.com/app/frohoff/ysoserial/file/2167893690/issues/source?bid=3161887&fileBranchId=3161890#l27)

File: src/main/java/ysoserial/secmgr/ExecCheckingSecurityManager.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+	@Override
+	public void checkPermission(final Permission perm, final Object context) { }
```
Comment (frohoff) [inline src/main/java/ysoserial/secmgr/ExecCheckingSecurityManager.java]: Codacy Issue found: [Document empty method body](https://www.codacy.com/app/frohoff/ysoserial/file/2167893690/issues/source?bid=3161887&fileBranchId=3161890#l30)

File: src/main/java/ysoserial/payloads/util/PayloadRunner.java [hunk 6]
```diff
@@ -1,40 +1,41 @@
...
-		try {	
-			System.out.println("deserializing payload");			
-			final Object objAfter = deserialize(serialized);			
+
+				return Serializer.serialize(objBefore);
+		}});
+
+		try {
+			System.out.println("deserializing payload");
+			final Object objAfter = Deserializer.deserialize(serialized);
```
Comment (frohoff) [inline src/main/java/ysoserial/payloads/util/PayloadRunner.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Deserializer.deserialize' due to existing static import 'ysoserial.Deserializer.deserialize'](https://www.codacy.com/app/frohoff/ysoserial/file/2167893687/issues/source?bid=3161887&fileBranchId=3161890#l34)

File: src/main/java/ysoserial/payloads/util/PayloadRunner.java [hunk 7]
```diff
@@ -1,40 +1,41 @@
...
 
-import static ysoserial.payloads.util.Serializables.deserialize;
-import static ysoserial.payloads.util.Serializables.serialize;
-
 import java.util.concurrent.Callable;
 
-import ysoserial.ExecBlockingSecurityManager;
+import ysoserial.Deserializer;
+import ysoserial.Serializer;
+import static ysoserial.Deserializer.deserialize;
```
Comment (frohoff) [inline src/main/java/ysoserial/payloads/util/PayloadRunner.java]: Codacy Issue found: [Avoid unused imports such as 'ysoserial.Deserializer.deserialize'](https://www.codacy.com/app/frohoff/ysoserial/file/2167893687/issues/source?bid=3161887&fileBranchId=3161890#l7)

File: src/main/java/ysoserial/payloads/util/PayloadRunner.java [hunk 8]
```diff
@@ -1,40 +1,41 @@
...
+import static ysoserial.Serializer.serialize;
```
Comment (frohoff) [inline src/main/java/ysoserial/payloads/util/PayloadRunner.java]: Codacy Issue found: [Avoid unused imports such as 'ysoserial.Serializer.serialize'](https://www.codacy.com/app/frohoff/ysoserial/file/2167893687/issues/source?bid=3161887&fileBranchId=3161890#l8)

File: src/main/java/ysoserial/payloads/util/PayloadRunner.java [hunk 9]
```diff
@@ -1,40 +1,41 @@
...
 				System.out.println("serializing payload");
-				
-				return serialize(objBefore);
-		}});			
-			
-		try {	
-			System.out.println("deserializing payload");			
-			final Object objAfter = deserialize(serialized);			
+
+				return Serializer.serialize(objBefore);
```
Comment (frohoff) [inline src/main/java/ysoserial/payloads/util/PayloadRunner.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Serializer.serialize' due to existing static import 'ysoserial.Serializer.serialize'](https://www.codacy.com/app/frohoff/ysoserial/file/2167893687/issues/source?bid=3161887&fileBranchId=3161890#l29)

File: src/main/java/ysoserial/Deserializer.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+	}
+
+	public static Object deserialize(final InputStream in) throws ClassNotFoundException, IOException {
+		final ObjectInputStream objIn = new ObjectInputStream(in);
+		return objIn.readObject();
+	}
+
+	public static void main(String[] args) throws ClassNotFoundException, IOException {
+		final InputStream in = args.length == 0 ? System.in : new FileInputStream(new File(args[0]));
+		Object object = deserialize(in);
```
Comment (frohoff) [inline src/main/java/ysoserial/Deserializer.java]: Codacy Issue found: [Avoid unused local variables such as 'object'.](https://www.codacy.com/app/frohoff/ysoserial/file/2167893672/issues/source?bid=3161887&fileBranchId=3161890#l32)

**Comment by frohoff** [comment]
I'm still leaning away from trying to do any gadget/chain-specific detection methods as they really are a poor quality indicator of vulnerability, and this concept will not scale very well as more gadgets are found (counting open PRs, we're up to ~20 gadgets).

I agree with the advice given by @pwntester and @cschneider4711 in their recent talk for stuff like this:
https://www.rsaconference.com/writable/presentations/file_upload/asd-f03-serial-killer-silently-pwning-your-java-endpoints.pdf

> Recommendation: Adjust active scanning payloads to not rely on specific gadgets - better use a generic delay introduction such as "SerialDoS" (by Wouter Coekaerts), which is only HashSet based.

I will probably close this, but regardless, thanks for the contributions and thoughtful discussion, and please keep educating people about the perils of unsafe deserialization.

**Comment by frohoff** [comment]
Closing. Thanks.

**Pull request closed without merging**